Mathlib/Data/Multiset/DershowitzManna.lean

```diff
@@ -0,0 +1,294 @@
+  induction M using Multiset.induction_on with
+  | empty =>
+    constructor
+    intro y y_lt
+    absurd y_lt
+    rintro ⟨X, Y, a, _, nonsense, _⟩
+    have contra : a ∈ (0 : Multiset α):= by
+      simp_all
+    contradiction
+  | cons _ _ ih =>
+    apply acc_cons_of_acc_of_lt
+    · apply wf_el
+      simp_all
+    · apply ih
+      intros
+      apply wf_el
+      simp_all
+
+/-- Over a well-founded order, `OneStep` is well-founded. -/
+private lemma isDershowitzMannaLT_singleton_wf (wf_lt : WellFoundedLT α) :
+    WellFounded (OneStep : Multiset α → Multiset α → Prop) := by
+  constructor
+  intros a
+  apply acc_of_acc_lt
+  intros x _
+  apply wf_lt.induction x
+  intros y h
+  apply Acc.intro y
+  assumption
+
+private lemma transGen_oneStep_of_isDershowitzMannaLT :
+    IsDershowitzMannaLT M N → TransGen OneStep M N := by
+  classical
+  rintro ⟨X, Y, Z, Z_not_empty, MXY, NXZ, h⟩
+  revert X Y M N
+  induction Z using strongInductionOn
+  case ih Z IH =>
+    rintro M N X Y rfl rfl Y_lt_Z
+    cases em (card Z = 0)
+    · simp_all
+    cases em (card Z = 1)
+    case inl hyp' hyp =>
+      rw [card_eq_one] at hyp
+      obtain ⟨z, rfl⟩ := hyp
+      apply TransGen.single
+      refine ⟨X, Y, z, rfl, rfl, ?_⟩
+      simp only [mem_singleton, exists_eq_left] at Y_lt_Z
+      exact Y_lt_Z
+    case inr hyp' hyp =>
+      have : ∃ a, a ∈ Z := by
+        rw [← Z.card_pos_iff_exists_mem]
+        cases Z_empty : card Z
+        tauto
+        simp
+      rcases this with ⟨z,z_in_Z⟩
+      let newZ := Z.erase z
+      have newZ_nonEmpty : newZ ≠ 0 := by
+        simp [newZ, ← sub_singleton, tsub_eq_zero_iff_le]
+        aesop
+      have newZ_sub_Z : newZ < Z := by simp (config := {zetaDelta := true}); exact z_in_Z
+      let f : α → Multiset α := fun z => Y.filter (fun y => y < z)
+      let N' := X + newZ + f z
+      apply @transitive_transGen _ _ _ N'
+      -- step from `N'` to `M`
+      · apply IH newZ newZ_sub_Z newZ_nonEmpty
+        · change _ = (X + f z) + (Y - f z)
+          ext a
+          have count_lt := count_le_of_le a (filter_le (fun y => y < z) Y)
+          simp_all only [empty_eq_zero, ne_eq, card_eq_zero, not_false_eq_true,
+            count_add, count_sub]
+          let y := count a Y
+          let x := count a X
+          let fz := count a (filter (fun x => x < z) Y)
+          change x + y = x + fz + (y - fz)
+          omega
+        · unfold N'
+          rw [add_assoc, add_assoc, add_comm newZ (f z)]
+        · intro y y_in
+          let Y_lt_Z := Y_lt_Z y
+          have y_in_Y : y ∈ Y := by aesop
+          let Y_lt_Z := Y_lt_Z y_in_Y
+          rcases Y_lt_Z with ⟨t, t_in_Z, y_lt_t⟩
+          use t
+          constructor
+          · by_cases t_in_newZ : t ∈ newZ
+            · exact t_in_newZ
+            · exfalso
+              have : t = z := by
+                have : Z = newZ + {z} := by
+                  unfold newZ
```

Comment:
```suggestion
```
This `unfold` is not needed, I suspect many other `unfold`s are not either, please check!

Reply:
Yes, you are right! I replaced all the other replaceable `unfold ..` as well ;-)